e2e/test/ConnectionStatusChangeHandlerTests.cs

```diff
@@ -0,0 +1,174 @@
+    using Microsoft.Azure.Devices.Client.Exceptions;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using System;
+    using System.Diagnostics.Tracing;
+    using System.Threading.Tasks;
+
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    public class ConnectionStatusChangeHandlerTests : IDisposable
+    {
+        private readonly string DevicePrefix = $"E2E_{nameof(ConnectionStatusChangeHandlerTests)}";
+        private readonly string ModulePrefix = $"E2E_{nameof(ConnectionStatusChangeHandlerTests)}";
+        private static TestLogging _log = TestLogging.GetInstance();
+
+        private readonly ConsoleEventListener _listener;
+
+        public ConnectionStatusChangeHandlerTests()
+        {
+            _listener = TestConfig.StartEventListener();
+        }
+
+        [TestMethod]
+        public async Task DeviceClient_DeviceDeleted_Gives_ConnectionStatus_DeviceDisabled()
+        {
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix + "_DeviceClient").ConfigureAwait(false);
+            string deviceConnectionString = testDevice.ConnectionString;
+
+            var config = new Configuration.IoTHub.DeviceConnectionStringParser(deviceConnectionString);
+            string deviceId = config.DeviceID;
+
+            ConnectionStatus? status = null;
+            ConnectionStatusChangeReason? statusChangeReason = null;
+            bool deviceDisabledReceived = false;
+
+            using (DeviceClient deviceClient =  DeviceClient.CreateFromConnectionString(deviceConnectionString, Client.TransportType.Amqp_Tcp_Only))
+            {
+                ConnectionStatusChangesHandler statusChangeHandler = (s, r) =>
+                {
+                    if (r == ConnectionStatusChangeReason.Device_Disabled)
+                    {
+                        status = s;
+                        statusChangeReason = r;
+                        deviceDisabledReceived = true;
+                    }
+                };
+
+                deviceClient.SetConnectionStatusChangesHandler(statusChangeHandler);
+                _log.WriteLine($"Created {nameof(DeviceClient)} ID={TestLogging.IdOf(deviceClient)}");
+
+                Console.WriteLine("DeviceClient OpenAsync.");
+                await deviceClient.OpenAsync().ConfigureAwait(false);
+
+                // Receiving the module twin should succeed right now.
+                Console.WriteLine("ModuleClient GetTwinAsync.");
+                var twin = await deviceClient.GetTwinAsync().ConfigureAwait(false);
+                Assert.IsNotNull(twin);
+
+                // Delete the device in IoT Hub. This should trigger the ConnectionStatusChangesHandler.
+                using (RegistryManager registryManager = RegistryManager.CreateFromConnectionString(Configuration.IoTHub.ConnectionString))
+                {
+                    await registryManager.RemoveDeviceAsync(deviceId).ConfigureAwait(false);
+                }
```

Comment:
Can you also add a test for disabling device?

Reply:
Yes, will do.

Disabling a device doesn't trigger an AMQP connection closure for ModuleClient somehow. Not entirely sure why so skipping that test for now.